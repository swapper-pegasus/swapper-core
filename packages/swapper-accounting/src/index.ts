import { getBalances } from './wallet/wallet'
import { WalletTokenBalanceData } from './types'

export { getBalances, WalletTokenBalanceData }
