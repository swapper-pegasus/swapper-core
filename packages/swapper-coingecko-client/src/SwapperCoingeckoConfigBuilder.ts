import qs from 'qs'

import {
  RequestConfigBuilder,
  RequestConfig,
  HttpMethod,
  Params
} from './http/HttpClientInterface'

type Data = Params;

export const THRESHOLD_AVOID_REQUEST_URL_TOO_LARGE = 4096

export default class SwapperCoingeckoConfigBuilder implements RequestConfigBuilder {
  private baseUrl: string
  private headers: object

  constructor ({ baseUrl }: {baseUrl: string;}) {
    this.baseUrl = baseUrl
    this.headers = {}
  }

  public async build (
    method: HttpMethod,
    path: string,
    params: Data,
    options?: { responseType: 'arraybuffer' }
  ) {
    const requestConfig: RequestConfig = {
      method,
      headers: this.headers,
      url: `${this.baseUrl}${path}`,
      ...(options || {})
    }

    switch (method) {
      case 'get': {
        const requestUrl = this.buildRequestUrl(path, params)
        if (requestUrl.length > THRESHOLD_AVOID_REQUEST_URL_TOO_LARGE) {
          return {
            ...requestConfig,
            method: 'post' as const,
            headers: this.headers,
            data: this.buildData(params)
          }
        }
        return {
          ...requestConfig,
          url: requestUrl
        }
      }
      case 'post': {
        return {
          ...requestConfig,
          data: this.buildData(params)
        }
      }
      case 'put': {
        return {
          ...requestConfig,
          data: this.buildData(params)
        }
      }
      case 'delete': {
        const requestUrl = this.buildRequestUrl(
          path,
          this.buildData(params)
        )
        return {
          ...requestConfig,
          url: requestUrl
        }
      }
      default: {
        throw new Error(`${method} method is not supported`)
      }
    }
  }

  private buildRequestUrl (path: string, params: Data): string {
    return `${this.baseUrl}${path}?${qs.stringify(params)}`
  }

  private buildData<T extends Data> (params: T) {
    return params
  }
}
