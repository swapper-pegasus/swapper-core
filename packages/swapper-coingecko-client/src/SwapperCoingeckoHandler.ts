import {
  ErrorResponse,
  HttpClientError,
  Response,
  ResponseHandler,
  ErrorCoingecko
} from './http/HttpClientInterface'

export default class SwapperCoingeckoHandler implements ResponseHandler {
  handle<T> (response: Promise<Response<T>>): Promise<T> {
    return response.then(
      (res) => this.handleSuccessResponse<T>(res),
      (error) => this.handleErrorResponse(error)
    )
  }

  private handleSuccessResponse<T> (response: Response<T>): T {
    return response.data
  }

  private handleErrorResponse (
    error: HttpClientError<ErrorResponse<ErrorCoingecko>>
  ): never {
    if (error.response?.data?.error) {
      throw new Error(error.response?.data?.error?.message)
    }
    throw new Error(error.message)
  }
}
