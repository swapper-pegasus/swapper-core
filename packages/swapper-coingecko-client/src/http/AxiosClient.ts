/* eslint-disable @typescript-eslint/no-explicit-any */
import Axios from 'axios'
import {
  HttpClient,
  RequestConfigBuilder,
  RequestConfig,
  ResponseHandler
} from './HttpClientInterface'

export default class AxiosClient implements HttpClient {
  private responseHandler: ResponseHandler
  private requestConfigBuilder: RequestConfigBuilder

  constructor ({
    responseHandler,
    requestConfigBuilder
  }: {
    responseHandler: ResponseHandler;
    requestConfigBuilder: RequestConfigBuilder;
  }) {
    this.responseHandler = responseHandler
    this.requestConfigBuilder = requestConfigBuilder
  }

  public async get<T extends object> (path: string, params: any) {
    const requestConfig = await this.requestConfigBuilder.build(
      'get',
      path,
      params
    )
    return (await this.sendRequest(requestConfig)) as Promise<T>
  }

  public async post<T extends object> (path: string, params: any) {
    const requestConfig = await this.requestConfigBuilder.build(
      'post',
      path,
      params
    )
    return (await this.sendRequest(requestConfig)) as Promise<T>
  }

  public async put<T extends object> (path: string, params: any) {
    const requestConfig = await this.requestConfigBuilder.build(
      'put',
      path,
      params
    )
    return (await this.sendRequest(requestConfig)) as Promise<T>
  }

  public async delete<T extends object> (path: string, params: any) {
    const requestConfig = await this.requestConfigBuilder.build(
      'delete',
      path,
      params
    )
    return (await this.sendRequest(requestConfig)) as Promise<T>
  }

  private sendRequest (requestConfig: RequestConfig) {
    return this.responseHandler.handle(
      Axios({
        ...requestConfig,
        maxBodyLength: Infinity,
        maxContentLength: Infinity
      })
    )
  }
}
