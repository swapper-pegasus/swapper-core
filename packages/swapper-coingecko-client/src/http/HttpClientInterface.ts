/* eslint-disable @typescript-eslint/no-explicit-any */
export interface HttpClient {
  get: <T extends object>(path: string, params: object) => Promise<T>;
  post: <T extends object>(path: string, params: object) => Promise<T>;
  put: <T extends object>(path: string, params: object) => Promise<T>;
  delete: <T extends object>(path: string, params: object) => Promise<T>;
}

export type ErrorResponse<T = any> = {
  data: T;
  status: number;
  statusText: string;
  headers: any;
};

export type ErrorCoingecko = {
  error: string
}

export type Response<T = any> = {
  data: T;
  headers: any;
};

export type HttpMethod = 'get' | 'post' | 'put' | 'delete';
export type Params = { [key: string]: unknown };

export interface HttpClientError<T = ErrorResponse> extends Error {
  response?: T;
}
export interface ResponseHandler {
  handle: <T = any>(response: Promise<Response<T>>) => Promise<T>;
}

export type RequestConfig = {
  method: HttpMethod;
  url: string;
  headers: any;
  httpsAgent?: any;
  data?: any;
};

export interface RequestConfigBuilder {
  build: (
    method: HttpMethod,
    path: string,
    params: Params,
    options?: { responseType: 'arraybuffer' }
  ) => Promise<RequestConfig>;
}
