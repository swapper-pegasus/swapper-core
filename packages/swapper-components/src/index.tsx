export * from './components/SwapComponent'
export * from './components/SearchCurrencyComponent'
export * from './components/List'
export * from './components/ModalKeystorePassword'
export * from './components/ModalInfo'
export * from './components/ModalDeposit'
export { ModalWithdrawal } from './components/ModalWithdrawal'
