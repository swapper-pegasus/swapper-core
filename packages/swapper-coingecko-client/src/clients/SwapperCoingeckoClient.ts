import { HttpClient, AxiosClient as DefaultHttpClient } from '../http'
import ISwapperCoingeckoClient from './interfaces/ISwapperCoingeckoClient'
import SwapperCoingeckoConfigBuilder from '../SwapperCoingeckoConfigBuilder'
import SwapperCoingeckoHandler from '../SwapperCoingeckoHandler'
import {
  // RESPONSE
  ResponseGetTokenPrice,
  ResponseGetTokens,
  // REQUEST
  RequestGetTokenPrice,
  // COMMON
  Token
} from './types'

export const USD_SYMBOL = 'usd'

export default class SwapperCoingeckoClient implements ISwapperCoingeckoClient {
  private client: HttpClient
  private cache: Array<Token>

  constructor (
    baseUrl: string,
    requestConfigBuilder = new SwapperCoingeckoConfigBuilder({ baseUrl: baseUrl }),
    responseHandler = new SwapperCoingeckoHandler(),
    client = new DefaultHttpClient({
      responseHandler,
      requestConfigBuilder
    })
  ) {
    this.client = client
  }

  public async getTokensPriceInUsd (requestPayload: RequestGetTokenPrice): Promise<ResponseGetTokenPrice> {
    const allTokens = await this.getTokens()
    const idSelectedTokens: Array<string | undefined> = requestPayload.tokens.map((symbolSelectedToken) => {
      const selectedToken: Token | undefined = allTokens.find((token) => token.symbol === symbolSelectedToken)
      return selectedToken?.id
    })
    const params = {
      ids: idSelectedTokens.join(','),
      vs_currencies: USD_SYMBOL
    }
    return this.client.get('/simple/price', params)
  }

  public async getTokens (): Promise<ResponseGetTokens> {
    if (!this.cache) {
      this.cache = await this.client.get('/coins/list', {})
    }
    return new Promise((resolve) => resolve(this.cache))
  }

  public async getTokenDataBySymbol (tokenSymbols: Array<string>): Promise<ResponseGetTokens> {
    const tokens = await this.getTokens()
    const selectedTokens = tokens.filter((token) => tokenSymbols.includes(token.symbol))
    return selectedTokens
  }
}
