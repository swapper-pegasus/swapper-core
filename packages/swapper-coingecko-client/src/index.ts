import SwapperCoingeckoClient from './clients/SwapperCoingeckoClient'

export * from './clients/types'

export {
  SwapperCoingeckoClient
}
