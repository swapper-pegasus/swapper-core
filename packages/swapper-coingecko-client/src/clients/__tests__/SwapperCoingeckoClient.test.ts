import { AxiosClient } from 'src/http'
import SwapperCoingeckoClient, { USD_SYMBOL } from '../SwapperCoingeckoClient'
import SwapperCoingeckoConfigBuilder from '../../SwapperCoingeckoConfigBuilder'
import SwapperCoingeckoHandler from '../../SwapperCoingeckoHandler'

class DefaultHttpClient {
  public get (): Promise<Array<object>> {
    return new Promise((resolve) => resolve([
      {
        id: 'tether',
        name: 'Thether',
        symbol: 'usdt'
      },
      {
        id: 'binance-usd',
        name: 'Binance USD',
        symbol: 'busd'
      }
    ]))
  }
}

const roundPrices = (tokens: object): Array<number> => {
  return Object.values(tokens).map(value => Math.round(value[USD_SYMBOL]))
}

describe('SwapperCoingeckoClient', () => {
  describe('getTokenPrice', () => {
    // It's hitting real endpoint
    it('success', async () => {
      const coingeckoClient: SwapperCoingeckoClient = new SwapperCoingeckoClient('https://api.coingecko.com/api/v3')
      const result = await coingeckoClient.getTokensPriceInUsd({ tokens: ['usdt', 'usdc', 'busd'] })
      expect(roundPrices(result)).toMatchSnapshot()
    })
    it('not supported token', async () => {
      const coingeckoClient: SwapperCoingeckoClient = new SwapperCoingeckoClient('https://api.coingecko.com/api/v3')
      const result = await coingeckoClient.getTokensPriceInUsd({ tokens: ['invent', 'usdc', 'busd'] })
      expect(roundPrices(result)).toMatchSnapshot()
    })
  })
  it('getTokens', async () => {
    // TODO: Better mock strategy for testing APIs
    const handler = new SwapperCoingeckoHandler()
    const builder = new SwapperCoingeckoConfigBuilder({ baseUrl: 'https://fake.api/api/v3' })
    const coingeckoClient: SwapperCoingeckoClient = new SwapperCoingeckoClient(
      'https://api.coingecko.com/api/v3',
      builder,
      handler,
      new DefaultHttpClient() as unknown as AxiosClient
    )
    const result = await coingeckoClient.getTokens()
    expect(result).toMatchSnapshot()
  })
})
