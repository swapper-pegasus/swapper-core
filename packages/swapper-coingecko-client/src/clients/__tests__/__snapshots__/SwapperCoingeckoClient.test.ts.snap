// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapperCoingeckoClient getTokenPrice not supported token 1`] = `
Array [
  1,
  1,
]
`;

exports[`SwapperCoingeckoClient getTokenPrice success 1`] = `
Array [
  1,
  1,
  1,
]
`;

exports[`SwapperCoingeckoClient getTokens 1`] = `
Array [
  Object {
    "id": "tether",
    "name": "Thether",
    "symbol": "usdt",
  },
  Object {
    "id": "binance-usd",
    "name": "Binance USD",
    "symbol": "busd",
  },
]
`;
